From a0bdc4e60f8a5e888422479d34006dce849f8144 Mon Sep 17 00:00:00 2001
From: bbhtt <bbhtt.zn0i8@slmail.me>
Date: Mon, 25 Mar 2024 09:20:30 +0530
Subject: [PATCH] clc: Use clang binary path to find clc_res_path

LLVM can't handle multi arch libdirs like `/usr/lib/$arch-triplet` and
returns a wrong clang_res_path `/usr/lib/lib/x86_64-linux-gnu/clang/17/include`
in the SDK. This is due to some issues with relative path and it doesn't
expose the proper functions for mesa to handle this.

So temporarily patch clang_res_path and pass the clang binary path to it
instead of libclang path, so that clang_res_path points to the correct
include directory: `/usr/lib/$arch-triplet/clang/17/include`
---
 src/compiler/clc/clc_helpers.cpp | 8 +++++++-
 1 file changed, 7 insertions(+), 1 deletion(-)

diff --git a/src/compiler/clc/clc_helpers.cpp b/src/compiler/clc/clc_helpers.cpp
index 530196b8328..0befa415be9 100644
--- a/src/compiler/clc/clc_helpers.cpp
+++ b/src/compiler/clc/clc_helpers.cpp
@@ -37,6 +37,7 @@
 #include <llvm/MC/TargetRegistry.h>
 #include <llvm/Target/TargetMachine.h>
 #include <llvm/Support/raw_ostream.h>
+#include <llvm/Support/Program.h>
 #include <llvm/Bitcode/BitcodeWriter.h>
 #include <llvm/Bitcode/BitcodeReader.h>
 #include <llvm-c/Core.h>
@@ -876,10 +877,15 @@ clc_compile_to_llvm_module(LLVMContext &llvm_ctx,
       return {};
    }
 
+    const auto clang_bin_path = (::llvm::sys::findProgramByName("clang").get());
+    if (clang_bin_path.empty()) {
+        clc_error(logger, "Couldn't find clang binary path.\n");
+    }
+
    // GetResourcePath is a way to retrive the actual libclang resource dir based on a given binary
    // or library.
    auto clang_res_path =
-      fs::path(Driver::GetResourcesPath(std::string(clang_path), CLANG_RESOURCE_DIR)) / "include";
+      fs::path(Driver::GetResourcesPath(std::string(clang_bin_path), CLANG_RESOURCE_DIR)) / "include";
    free(clang_path);
 
    c->getHeaderSearchOpts().UseBuiltinIncludes = true;
-- 
2.44.0

